sops-smoke-test/src/test/kotlin/cryptoservice/service/ContainerTest.kt

```diff
@@ -0,0 +1,85 @@
+import org.junit.jupiter.api.AfterAll
+import org.junit.jupiter.api.Assertions.assertEquals
+import org.junit.jupiter.api.Assertions.assertTrue
+import org.junit.jupiter.api.BeforeAll
+import org.junit.jupiter.api.Test
+import org.slf4j.LoggerFactory
+import org.springframework.http.MediaType
+import org.springframework.http.ProblemDetail
+import org.springframework.test.web.reactive.server.WebTestClient
+import org.testcontainers.containers.Container
+import org.testcontainers.containers.GenericContainer
+import org.testcontainers.containers.output.Slf4jLogConsumer
+
+class ContainerTest {
+    @Test
+    fun `call encrypt`() {
+        val address = "http://${cryptoServiceContainer.host}:${cryptoServiceContainer.getMappedPort(8080)}"
+
+        val expectedDetailRegex =
+            Regex(
+                "Exception message: Failed when encrypting RiSc with ID: some-id ",
+            )
+        val client = WebTestClient.bindToServer().baseUrl(address).build()
+        client
+            .post()
+            .uri("/encrypt")
+            .accept(MediaType.APPLICATION_JSON)
+            .contentType(MediaType.APPLICATION_JSON)
+            .bodyValue(
+                """
+                {
+                    "text": "Hello",
+                    "config": {"gcp_kms": null},
+                    "gcpAccessToken": "fake-token",
+                    "riScId": "some-id"
+                }
+                """.trimIndent(),
+            ).exchange()
+            .expectStatus()
+            .is5xxServerError
+            .expectBody(SimpleErrorResponse::class.java)
+            .value({ errorResponse ->
+                println("ACTUAL RESPONSE: ${errorResponse.body.detail}")
+                assertTrue(errorResponse.body.detail?.matches(expectedDetailRegex) ?: false, {
+                    """
+                    Unexpected error detail:
+                    Expected ${expectedDetailRegex.pattern}
+                    actual: ${errorResponse.body.detail}
+                    """.trimIndent()
+                })
+            })
+    }
+
+    @Test
+    fun `check that sops exists as a command in container`() {
+        val sopsResult: Container.ExecResult = cryptoServiceContainer.execInContainer("sops", "--version")
+        val exitCode: Int = sopsResult.exitCode
+        assertEquals(exitCode, 0)
```

Comment:
Det første argumentet til `assertEquals` er hva vi forventer, mens det andre er den faktiske verdien koden har produsert, så må gjøre følgende.

```suggestion
        val exitCode: Int = sopsResult.exitCode
        assertEquals(0, exitCode)
```

Koden blir enklere om vi gjøre:
```suggestion
        assertEquals(0, sopsResult.exitCode)
```

Reply:
Grunnen til at testen er i en egen mappe er for å ha et skille mellom hva som er applikasjonskode og hva som er "test-av-ferdig-container"-kode. Så slipper en miksing og triksing mellom tester som er skrudd av og på avhengig av om en kjører lokalt eller på CI. Og en slipper den rekursive avhengigheten der både kode og "noe utenfor" må bygges for at testen som ligger sammen med koden skal kunne kjøre. Så tydelig separasjon mellom applikasjon + applikasjontestkode, og containertestkode.